process explorer/backend/ProcessExplorer/Processes/IProcessGenerator.cs

```diff
@@ -0,0 +1,245 @@
+        {
+
+        }
+        public int? GetParentId(Process process)
+        {
+            int parentPid = 0;
+            using (ManagementObject mo = new ManagementObject("win32_process.handle='" + process.Id.ToString() + "'"))
+            {
+                mo.Get();
+                parentPid = Convert.ToInt32(mo["ParentProcessId"]);
+            }
+            return parentPid;
+        }
+
+        public float GetMemoryUsage(Process process)
+        {
+            int memsize = 0;
+            PerformanceCounter PC = new PerformanceCounter();
+            PC.CategoryName = "Process";
+            PC.CounterName = "Working Set - Private";
+            PC.InstanceName = process.ProcessName;
+            memsize = Convert.ToInt32(PC.NextValue()) / (int)(1024) / (int)1024;
+            PC.Close();
+            PC.Dispose();
+            return (float)(memsize / GetTotalMemoryInMB()) * 100;
+        }
+
+        private static double GetTotalMemoryInMB()
+        {
+            var gcMemoryInfo = GC.GetGCMemoryInfo();
+            var installedMemory = gcMemoryInfo.TotalAvailableMemoryBytes;
+            return (double)installedMemory / 1048576.0;
+        }
+
+        public float GetCPUUsage(Process process)
+        {
+            var cpu = new PerformanceCounter("Process", "% Processor Time", process.ProcessName);
+            return cpu.NextValue() * 100;
+        }
+
+        public SynchronizedCollection<ProcessInfoDto> GetChildProcesses(Process process)
+        {
+            SynchronizedCollection<ProcessInfoDto> children = new SynchronizedCollection<ProcessInfoDto>();
+            ManagementObjectSearcher mos = new ManagementObjectSearcher(string.Format("Select * From Win32_Process Where ParentProcessID={0}", process.Id));
+            foreach (ManagementObject mo in mos.Get())
+            {
+                children.Add(new ProcessInfo(Process.GetProcessById(Convert.ToInt32(mo["ProcessID"])), this).Data);
+            }
+
+            return children;
+        }
+        public void WatchProcesses(SynchronizedCollection<int> processes)
+        {
+            this.processes = processes;
+            // create the watcher and start to listen
+            try
+            {
+                string WmiQuery;
+                ManagementEventWatcher Watcher;
+                ManagementScope Scope;
+
+                Scope = new ManagementScope(@"\\.\root\CIMV2");
+                Scope.Connect();
+
+                WmiQuery = "Select * From __InstanceCreationEvent Within 1 " +
+                "Where TargetInstance ISA 'Win32_Process' ";
+
+                Watcher = new ManagementEventWatcher(Scope, new EventQuery(WmiQuery));
+                Watcher.EventArrived += new EventArrivedEventHandler(WmiEventHandler);
+                Watcher.Start();
+                //Watcher.Stop();
+            }
+            catch (Exception)
+            {
+                throw new ApplicationException("Error while reading the processes.");
```

Comment:
https://docs.microsoft.com/en-us/dotnet/api/system.applicationexception?view=net-6.0
"You should not throw an [ApplicationException](https://docs.microsoft.com/en-us/dotnet/api/system.applicationexception?view=net-6.0) exception in your code,"